libbroker/broker/hub.test.cc

```diff
@@ -0,0 +1,94 @@
+#include "broker/hub.hh"
+
+#include "broker/broker-test.test.hh"
+#include "broker/builder.hh"
+#include "broker/publisher.hh"
+
+#include <chrono>
+
+using namespace std::literals;
+
+TEST("hubs do not receive their own messages") {
+  broker::endpoint ep;
+  auto uut = ep.make_hub({"/foo/bar"});
+  uut.publish("/foo/bar", broker::list_builder{}.add(1).add(2).add(3));
+  auto msg = uut.get(150ms);
+  CHECK_EQ(msg, nullptr);
+}
+
+TEST("hubs receive messages from other hubs") {
+  broker::endpoint ep;
+  auto uut1 = ep.make_hub({"/foo/bar"});
+  auto uut2 = ep.make_hub({"/foo/bar"});
+  uut1.publish("/foo/bar", broker::list_builder{}.add(1).add(2).add(3));
+  auto msg = uut2.get(150ms);
+  if (CHECK_NE(msg, nullptr)) {
+    CHECK_EQ(msg->topic(), "/foo/bar");
+    if (auto val = msg->value(); CHECK(val.is_list())) {
+      CHECK_EQ(broker::to_string(val), "(1, 2, 3)");
+    }
+  }
+}
+
+TEST("hubs can change their subscriptions") {
+  broker::endpoint ep;
+  auto uut1 = ep.make_hub({"/foo/baz"});
+  auto uut2 = ep.make_hub({"/foo/bar"});
+  // Send to /foo/baz: should not be received by uut2.
+  uut1.publish("/foo/baz", broker::list_builder{}.add(1).add(2).add(3));
+  if (auto msg = uut2.get(150ms); !CHECK_EQ(msg, nullptr)) {
+    return;
+  }
+  // Subscribe to /foo/baz on uut2: should receive the message.
+  uut2.subscribe("/foo/baz", true);
+  uut1.publish("/foo/baz", broker::list_builder{}.add(4).add(5).add(6));
+  if (auto msg = uut2.get(150ms); CHECK_NE(msg, nullptr)) {
+    CHECK_EQ(msg->topic(), "/foo/baz");
+    if (auto val = msg->value(); CHECK(val.is_list())) {
+      CHECK_EQ(broker::to_string(val), "(4, 5, 6)");
+    }
+  }
+  // Unsubscribe from /foo/baz on uut2: should not receive the message anymore.
+  uut2.unsubscribe("/foo/baz", true);
+  uut1.publish("/foo/baz", broker::list_builder{}.add(7).add(8).add(9));
+  if (auto msg = uut2.get(150ms); !CHECK_EQ(msg, nullptr)) {
+    return;
+  }
+}
+
+TEST("hubs receive messages from publishers") {
+  broker::endpoint ep;
+  auto uut = ep.make_hub({"/foo/bar"});
+  auto pub = ep.make_publisher(broker::topic{"/foo/bar"});
+  pub.publish(broker::list_builder{}.add(1).add(2).add(3));
+  auto msg = uut.get(150ms);
+  if (CHECK_NE(msg, nullptr)) {
+    CHECK_EQ(msg->topic(), "/foo/bar");
+    if (auto val = msg->value(); CHECK(val.is_list())) {
+      CHECK_EQ(broker::to_string(val), "(1, 2, 3)");
+    }
+  }
+}
+
+TEST("hubs receive messages that are published on the endpoint") {
+  broker::endpoint ep;
+  auto uut = ep.make_hub({"/foo/bar"});
+  ep.publish(broker::topic{"/foo/bar"},
+             broker::list_builder{}.add(1).add(2).add(3).build());
+  auto msg = uut.get(150ms);
+  if (CHECK_NE(msg, nullptr)) {
+    CHECK_EQ(msg->topic(), "/foo/bar");
+    if (auto val = msg->value(); CHECK(val.is_list())) {
+      CHECK_EQ(broker::to_string(val), "(1, 2, 3)");
+    }
+  }
+}
+
+TEST("subscribers do not receive messages from hubs") {
```

Comment:
I'm not completely sure what the intuition for this one is.

Is it because `std::endpoint.publish()` or `publisher::publish()` would need more special cases? Given `endpoint::publish()` and `publisher::publish()` reach hubs, it seems reasonable to have publish on hubs reach bare subscribers, but I presume you had reasons to do it this way?

I believe this means that `broker/Manager.cc` needs to switch to use hub for basic pub/sub. Which is certainly okay.

Reply:
The model for subscribers is simply that they don't receive messages that are published locally. Neither from a `publisher`, nor when using `endpoint::publish`, nor when using a `hub`.

These changes should be 100% transparent to Zeek.

Reply:
I guess that is consistent, though it does mean the subscriber isn't really useful for Zeek's pub/sub and should go through hubs completely. That's okay!